### PR TITLE
MODEMAIL-21 Implement an endpoint for getting all emails for metrics

### PR DESCRIPTION
**Pull request by OleksiiMaksymov**
**Purpose**
Getting all emails for metrics
Resolves: [MODEMAIL-21](https://issues.folio.org/browse/MODEMAIL-21)

**Approach**
1. Persist all emails in the database for some time (specified or default period on the basis of which will be carried out cleaning or archiving emails).
2. Handle SMTP server response status (delivered, failed, and other failures associated with an error within the module) and reasons for sending failure
3. Implement an endpoint to get processed emails (different interface)

### CODE REVIEW

File: ramls/delayedTasks.raml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+types:
+  errors: !include raml-util/schemas/errors.schema
+
+traits:
+  validate: !include raml-util/traits/validation.raml
+  language: !include raml-util/traits/language.raml
+  pageable:  !include raml-util/traits/pageable.raml
+  searchable: !include raml-util/traits/searchable.raml
+
+resourceTypes:
```
Comment (opochepam) [inline ramls/delayedTasks.raml]: unused resource types

File: ramls/delayedTasks.raml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+          type: string
+          required: false
+      description: "delete expired email messages"
+      responses:
+        204:
+          description: "success"
+          body:
+            text/plain:
+              example: "No Content"
+        400:
```
Comment (opochepam) [inline ramls/delayedTasks.raml]: unused response

File: ramls/email_entity_collections.json [hunk 3]
```diff
@@ -0,0 +1,21 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
```
Comment (opochepam) [inline ramls/email_entity_collections.json]: missed required section

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+public class DelayedTasksAPI extends AbstractEmail implements DelayedTask {
+
+  public DelayedTasksAPI(Vertx vertx, String tenantId) {
+    super(vertx, tenantId);
+  }
+
+  @Override
+  public void deleteDelayedTaskExpiredMessages(String expirationDate, String status, Map<String, String> headers,
+                                               Handler<AsyncResult<Response>> resultHandler, Context context) {
+    try {
```
Comment (opochepam) [inline src/main/java/org/folio/rest/impl/DelayedTasksAPI.java]: you can get rid of try/catch staring with succeeded future and call checkExpirationDate in the compose lambda

File: src/main/java/org/folio/rest/impl/base/AbstractEmail.java [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+import io.vertx.core.http.HttpHeaders;
+import io.vertx.core.json.JsonObject;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+
+public abstract class AbstractEmail {
+
+  private static final String REQUEST_URL_TEMPLATE = "%s/%s?query=module==%s";
+  private static final String REQUEST_URI_PATH = "configurations/entries";
+  private static final String HTTP_HEADER_ACCEPT = HttpHeaders.ACCEPT.toString();
```
Comment (opochepam) [inline src/main/java/org/folio/rest/impl/base/AbstractEmail.java]: redundant constant

Comment (OleksiiMaksymov) [reply src/main/java/org/folio/rest/impl/base/AbstractEmail.java]: fix

File: src/main/java/org/folio/rest/impl/DelayedTasksAPI.java (hunk 4 above)
Comment (OleksiiMaksymov) [reply src/main/java/org/folio/rest/impl/DelayedTasksAPI.java]: fix

File: ramls/email_entity_collections.json (hunk 3 above)
Comment (OleksiiMaksymov) [reply ramls/email_entity_collections.json]: fix

File: ramls/delayedTasks.raml (hunk 2 above)
Comment (OleksiiMaksymov) [reply ramls/delayedTasks.raml]: fix

File: ramls/delayedTasks.raml (hunk 1 above)
Comment (OleksiiMaksymov) [reply ramls/delayedTasks.raml]: fix

File: src/test/java/org/folio/rest/impl/base/AbstractAPITest.java [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+        } catch (MessagingException ex) {
+          logger.debug(ex);
+          throw throwAssertionFailedError(sender);
+        }
+      })
+      .findFirst()
+      .orElseThrow(() -> throwAssertionFailedError(sender));
+  }
+
+  private AssertionFailedError throwAssertionFailedError(String sender) {
```
Comment (opochepam) [inline src/test/java/org/folio/rest/impl/base/AbstractAPITest.java]: method's name does not reflect what it does correctly. it creates new error but does not throw it

Comment (OleksiiMaksymov) [reply src/test/java/org/folio/rest/impl/base/AbstractAPITest.java]: Updated

**Pull request merged**